New auxiliary view "grade_available_responses_view"

dmitry-viskov opened the pull request:
Auxiliary view `grade_available_responses_view` which is needed to display the staff grading area in the new Open Response Assessment tab in the Instructor Dashboard of LMS:
https://projects.invisionapp.com/share/YKA6S7VH8#/screens/216425777
![ora1](https://cloud.githubusercontent.com/assets/1876893/23087305/30535010-f584-11e6-893c-1cdf59229a2a.jpeg)
![ora2](https://cloud.githubusercontent.com/assets/1876893/23087310/4081dec0-f584-11e6-85a2-d6613b4fa759.jpeg)
This PR is the necessary part of this new feature: https://github.com/edx/edx-platform/pull/14372

edx-webhook commented:
Thanks for the pull request, @dmitry-viskov! I've created [OSPR-1677](https://openedx.atlassian.net/browse/OSPR-1677) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

dmitry-viskov commented:
This PR is the necessary part of this new feature: edx/edx-platform#14372

dmitry-viskov commented:
The main goal of this PR is to have ability in the LMS to do such things:
```
fragment = xblock.render('grade_available_responses_view')
```

cahrens commented:
@dmitry-viskov is there a sandbox where we can try this out?

dmitry-viskov commented:
@cahrens I may only give you the access to our test stand where this feature is already deployed. will that do?

efischer19 commented on openassessment/templates/openassessmentblock/oa_grade_available_responses.html:
```diff
@@ -0,0 +1,32 @@
+{% load i18n %}
```
Can you create a new folder, `instructor_dashboard`, under `openassessment/templates/openassessmentblock/` and move this file there? That seems to fit with the rest of the template organization a bit more cleanly

dmitry-viskov replied:
done

efischer19 commented on openassessment/templates/openassessmentblock/oa_grade_available_responses.html:
```diff
@@ -0,0 +1,32 @@
+            </div>
+            {% else %}
+                <div class="openassessment__staff-area-unavailable">
+                {% trans "Grade Available Responses for the chosen assessment is unavailable. This item is not configured for Staff Assessment." %}
```
Terminology note: `assessment` tends to refer specifically to the result of a reviewer assessing a submission in the context of ORA code. Can you change this line to `Grade Available Responses is unavailable. This item is not configured for Staff Assessment.`?

dmitry-viskov replied:
changed this

efischer19 commented on openassessment/xblock/static/js/src/lms/oa_base.js:
```diff
@@ -384,5 +384,15 @@ function OpenAssessmentBlock(runtime, element, data) {
 
+/* XBlock JavaScript entry point for OpenAssessmentXBlock. */
+/* jshint unused:false */
+function GradeAvailableResponsesBlock(runtime, element, data) {
```
Would `StaffAssessmentBlock` be a better name for this?

dmitry-viskov replied:
Changed this

efischer19 commented on openassessment/xblock/test/test_openassessment.py:
```diff
@@ -59,6 +59,33 @@ def test_load_student_view(self, xblock):
         self.assertIn("step--grade", grade_response.body)
 
+    @scenario('data/staff_grade_scenario.xml')
+    def test_load_grade_available_responses_view(self, xblock):
```
As above, I could see naming this `test_staff_assessment_view`. I'm not super picky which we end up with, as long it's consistent 😀

dmitry-viskov replied:
ok. Fixed

efischer19 commented on openassessment/xblock/test/test_openassessment.py:
```diff
@@ -59,6 +59,33 @@ def test_load_student_view(self, xblock):
+        self.assertIn("GradeAvailableResponsesBlock", body_html)
+        self.assertIn("openassessment__title", body_html)
+        self.assertNotIn("ui-staff__content", body_html)
+        self.assertIn("openassessment__staff-area-unavailable", body_html)
```
These two tests are identical outside of scenario and the final assert, yes? Let's combine them to avoid duplicating code.

```
@scenario(a)
def test_1(self):
    _staff_assessment_view_helper()
    self.assertNotIn(...)

@scenario(b)
def test_2(self):
    _staff_assessment_view_helper()
    self.assertIn(...)
```

dmitry-viskov replied:
ok. Moved common code to the `_staff_assessment_view_helper` method

cahrens commented on openassessment/templates/openassessmentblock/oa_grade_available_responses.html:
```diff
@@ -0,0 +1,32 @@
+    <div class="openassessment problem">
+        <div class="wrapper--grid">
+            {% if title %}
+                <h3 class="openassessment__title problem__header">{% trans title %}</h3>
```
Please make sure that the header levels are correct (I see h3 here and an h2 below), taking into account headers that are already present on the page.

@staubina can you advise what header level this should start with, given the work you are doing to combine the top-level h1 and h2?

staubina replied:
I missed this PR, I will take a look.  The initial set of header changes are merged now though.

staubina replied:
The header level is very hard to determine without seeing this in the platform.  @dmitry-viskov in the image above which text is the openassessment__title?

staubina replied:
Also with this header being conditional, you could run in to header level issues.  If it is not available there will be a gap.  Ensure the headers progress in a logical hierarchy.

staubina commented on openassessment/templates/openassessmentblock/oa_grade_available_responses.html:
```diff
@@ -0,0 +1,32 @@
+            <div class="openassessment__staff-area">
+                <div class="openassessment__staff-grading wrapper--staff-grading">
+                    <div class="staff-grading ui-staff">
+                        <h2 class="staff-grading__title">
```
Is this h2 below the h3?  Please review the order of the headers in this file.  This order is important for accessibility.

staubina commented:
The headers should be something like this, 

H2 - Instructor Dashboard
H3 - Censorship in the Libraries
H4 - Grade Available Responses
and so on

If you rebase on master you will pick up my changes and that might help.

dmitry-viskov commented:
@sarina thanks for review. I've made rebase and changed template a little. How it looks after my update:

![ora-headers](https://cloud.githubusercontent.com/assets/1876893/23926210/f4955d30-0923-11e7-9e61-18b157529fdc.jpg)

now looks better?

dmitry-viskov commented:
@efischer19 thanks for the review!

>Is this PR going to be updated with more changes as requested on edx/edx-platform#14562 (comment), or will that be a new PR?

yes. I've updated the PR: 1. fixed all your remarks, 2. created new auxiliary view `ora_blocks_listing_view` to show grid with all available course ORA blocks (grid is based on `backgrid.js`). I suggest that in LMS code this new view will be used in such way:

```
def _section_open_response_assessment(request, course, openassessment_blocks, access):
    """Provide data for the corresponding dashboard section """
    course_key = course.id

    ora_items = []
    parents = {}

    for block in openassessment_blocks:
        block_parent_id = unicode(block.parent)
        result_item_id = unicode(block.location)
        if block_parent_id not in parents:
            parents[block_parent_id] = modulestore().get_item(block.parent)

        ora_items.append({
            'id': result_item_id,
            'name': block.display_name,
            'parent_id': block_parent_id,
            'parent_name': parents[block_parent_id].display_name,
            'staff_assessment': 'staff-assessment' in block.assessment_steps,
            'url_base': reverse('xblock_view', args=[course.id, block.location, 'student_view']),
            'url_grade_available_responses': reverse('xblock_view', args=[course.id, block.location,
                                                                          'grade_available_responses_view']),
        })

    openassessment_block = openassessment_blocks[0]
    block, __ = get_module_by_usage_id(
        request, unicode(course_key), unicode(openassessment_block.location),
        disable_staff_debug_info=True, course=course
    )
    section_data = {
        'fragment': block.render('ora_blocks_listing_view', context={
            'ora_items': ora_items,
            'ora_item_view_enabled': settings.FEATURES.get('ENABLE_XBLOCK_VIEW_ENDPOINT', False)
        }),
        'section_key': 'open_response_assessment',
        'section_display_name': _('Open Responses'),
        'access': access,
        'course_id': unicode(course_key),
    }
    return section_data
```

staubina commented:
@dmitry-viskov 

> @staubina thanks for review. I've made rebase and changed template a little. How it looks after my update:

The issue still exists.  The Header "Instructor Dashboard" will be an H2, everything below that should be greater than H2.  

H2 - Instructor Dashboard
H3 - Open Responses
H4 - Open Response Assessment
H5 - Grade Available Responses

The greater question though is why are there so many headers right on top of each other?  It seems odd to stack 3 headers.

dmitry-viskov commented:
@staubina just for a note how the current headers are presented in the ORA block in LMS:

![screencapture](https://cloud.githubusercontent.com/assets/1876893/23948666/15400690-0995-11e7-8d35-38f797cfdd5f.png)

dmitry-viskov commented:
@staubina updated the PR again:

![ora_h](https://cloud.githubusercontent.com/assets/1876893/23948695/3a0e8988-0995-11e7-9b67-cd1d5eafc337.jpeg)

staubina commented:
>  just for a note how the current headers are presented in the ORA block in LMS:

Thank you for highlighting that issue.  When updating ORA2's headers we focused on the student experience.  The decision was made at the time to not adjust the staff assessment section while we focused on other a11y work.

staubina commented:
> @staubina updated the PR again:

The screenshot looks good.  I am looking at the code now.

staubina commented:
I focused on the header levels as I was asked.  From that standpoint the changes look good to me.  This is of course without being able to access a Sandbox.

cahrens commented:
@dmitry-viskov The header levels are technically correct now, but from a user perspective it's pretty confusing. Perhaps you can talk to @roderickmorales (our UX person) about how to simplify the presentation. 

@roderickmorales I am referring to

![image](https://cloud.githubusercontent.com/assets/484484/23951423/6178d2c0-0964-11e7-8594-7833633af6b3.png)

jbarciauskas commented:
Just to clarify, @cahrens - this isn't a blocker, based on our conversation earlier? Though it would be nice to see if there were a better solution?

cahrens commented:
@jbarciauskas Correct, not a blocker as long as the header levels are technically correct (which they seem to be from the image).

dmitry-viskov commented:
@efischer19 thanks for the review. 

Decision about creating grids based on backgridjs was discussed here: https://github.com/edx/edx-platform/pull/14372#issuecomment-278356112 with @andy-armstrong and @dsjen . They recommended me to use this library because it is already used in other edx projects.

To be honest I would not want to rewrite this functionality again because the current styles and JS are designed to use backgridjs. So change grid library will entail significant changes in the code (and will require significant amount of time on rework).

bjacobel commented on openassessment/xblock/openassessmentblock.py:
```diff
@@ -364,12 +364,87 @@ def student_view(self, context=None):
 
+        # backgrid.js v0.3.7
+        fragment.add_css_url(self.runtime.local_resource_url(self, "static/css/lib/backgrid.min.css"))
+        self.add_javascript_files(fragment, "static/js/lib/backgrid.min.js")
```
New JavaScript libraries should be installed through NPM and loaded from `node_modules`. The non-minifed version should be used (relying on our app's asset pipeline to do the minification ourselves for production, so that we can debug the unminified version in development)

Documentation on this can be found in our developer documentation for [Developing User Interfaces](http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/user_interface_development.html#adding-javascript-libraries) or our Confluence page for [adding a new JavaScript library](https://openedx.atlassian.net/wiki/display/FEDX/How+to+add+a+new+third+party+JavaScript+library+to+LMS+or+Studio).

dmitry-viskov replied:
@bjacobel I agree that it is necessary to add non-minified version for development (and I'll add it) but I disagree with you about using `Backgrid.js` version from NPM. Theoretically - yes, you're right. But in practice it doesn't work. This approach is working **only for tests** that are executed from the ORA projects. The links that you posted designed for LMS but here is XBlock.

When LMS install XBlock dependencies it uses only existed files (that are described in the [MANIFEST.in](https://github.com/edx/edx-ora2/blob/master/MANIFEST.in)). So you can't run `npm install` during install all these dependencies from pip. Of course I may add `Backgrid.js` to the minified version `openassessment-lms.min.js` by modifying this string:
https://github.com/edx/edx-ora2/blob/master/Makefile#L33
But it will work only for `settings.DEBUG == False`. In case of `settings.DEBUG == True` LMS requires non-minified versions: https://github.com/edx/edx-ora2/blob/master/openassessment/xblock/openassessmentblock.py#L373

Also I don't quite understand why this problem is so critical because as I may see there are a lot of dependencies that are used from the local folder (not from NPM). Examples:
https://github.com/edx/edx-ora2/tree/master/openassessment/xblock/static/js/lib
https://github.com/edx/edx-platform/tree/master/common/static/js/vendor

/cc @efischer19

efischer19 commented:
Apologies @dmitry-viskov - I had missed that original conversation. Backgrid is good to continue then, can you address the library installation concerns that @bjacobel just raised?

bjacobel commented:
@efischer19 @staubina @cahrens Wait, is Backbone just being added to this repo as of this PR?

We should not be creating dependencies on Backbone in any of our IDAs at this point. @edx/fedx-team has been communicating that Backbone will soon be deprecated since late last year.

dmitry-viskov commented:
@bjacobel backbone dependency is needed only for tests. In case of using ORA block within LMS Instructor dashboard backbone is taken from the LMS

cptvitamin commented:
hi @dmitry-viskov. I just wanted to point you to our Accessibility Developer Guidelines in case you weren't already familiar with them: http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/accessibility.html We will be expanding on this document substantially over the coming months, in an effort to make the guidance more comprehensive. However, the overarching requirement is that all new features added to the platform meet our WCAG 2 Level AA Conformance requirement.

Let me know if you have any questions.

dsjen commented:
I don't meant to muddy the waters here, but wanted to point that that the newest version of backgrid released a few weeks ago includes many accessibility, see https://github.com/cloudflare/backgrid/pull/626.  Leveraging that version may address some of the a11y concerns.

dmitry-viskov commented:
@cptvitamin @bjacobel I'm not quite sure that understand you. Does the current implementation of the particular feature in this PR have some particular accessibility issues? Can you describe them ? I don't see any accessibility problems here

cptvitamin commented:
Hi @dmitry-viskov, I only referenced our a11y requirements and made no claim to the accessibility of your feature or third party libraries used.  If this feature is ready for review, can you provide a link to an environment where it can be tested?  If you've done your own accessibility review and are prepared for an acceptance review, just let me know.

dmitry-viskov commented:
@cptvitamin 

>If this feature is ready for review

This feature is ready for review from the moment when I created the current PR :smile: 

>can you provide a link to an environment where it can be tested?

I'm not an edx team member and don't know where you test such features. I may suggest to deploy your our local environment using my branch or give access to our Credo test server where this feature is deployed. Another variant you may use the screenshots attached to this task (or these ones: https://github.com/edx/edx-platform/pull/14562)

>If you've done your own accessibility review

I'm not very familiar with accessibility guidelines or something like this. The design of this feature was discussed here: https://github.com/edx/edx-platform/pull/14372 and here the mockups: https://projects.invisionapp.com/share/YKA6S7VH8#/216097426_Course_-Instructor_-AppsAndTools
The current feature is implemented according these mockups. So If you have some offers to change font or  something like this - it is not a problem I may change it. Otherwise as from my opinion this feature is OK

cptvitamin commented:
@dmitry-viskov the accessibility of web application is based entirely on its implementation and interactivity, so testing a mockup will not work. I understand you may not be familiar with accessibility, but we do ask that you review the [Accessibility Guidelines](http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/accessibility.html) and perform your own [self-assessment](http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/accessibility.html#test-your-code-for-accessibility) based on the topics covered in that doc.

dmitry-viskov commented:
@cptvitamin myself-assessment is OK. if you have some suggestions i may change/implement them

efischer19 commented:
Hey @dmitry-viskov, sorry about all the confusion this afternoon. We've clarified things internally, can you provide the following items to continue with this review?

- the code changes I've already flagged above
- the latest version of backgrid, installed via npm [as suggested by @bjacobel earlier](https://github.com/edx/edx-ora2/pull/979#discussion_r106471789)
- access to your test server (I've tried getting your code running as-is, but the backgrid installation isn't playing well with my sandbox). Could @cptvitamin and I get that access please? We need to inspect live, running code.

Thanks, I'm really looking forward to this feature!

dmitry-viskov commented:
Thanks @efischer19 

>the code changes I've already flagged above

yes. Will be done soon.

>the latest version of backgrid

The last version of backgrid already used.

>installed via npm

I've posted my thoughts why this approach is not working: https://github.com/edx/edx-ora2/pull/979#discussion_r106561181

>access to your test server 

ok. no problem. I'll try to provide you access tomorrow. one note - our test server still based on eucalyptus version. But i hope that it will not be a obstacle

dmitry-viskov commented on openassessment/xblock/openassessmentblock.py:
```diff
@@ -97,21 +98,20 @@ def load(path):
+                          StudentTrainingMixin,
+                          LmsCompatibilityMixin,
+                          CourseItemsListingMixin,
+                          XBlock):
```
np! :+1:

dmitry-viskov commented on openassessment/xblock/openassessmentblock.py:
```diff
@@ -364,12 +364,87 @@ def student_view(self, context=None):
+        ora_item_view_enabled = context.get('ora_item_view_enabled', False) if context else False
+        context_dict = {
+            "ora_items": json.dumps(ora_items),
+            "ora_item_view_enabled": 1 if ora_item_view_enabled else 0
```
ok. changed this

dmitry-viskov commented on openassessment/xblock/static/js/spec/lms/oa_listing.js:
```diff
@@ -0,0 +1,281 @@
+        "url_base": "/student_view",
+        "id": "block-v1:SomeOrg+ORA101+2017+type@openassessment+block@40b4edfe60bc47d9ba037a3180c44e97"
+    }];
+
```
Created `oraCourseItems` variable: https://github.com/edx/edx-ora2/pull/979/files#diff-d629dbe5d581fab05aa0b5674c4aeef7R40

is that better?

dmitry-viskov commented on openassessment/xblock/static/js/spec/lms/oa_listing.js:
```diff
@@ -0,0 +1,281 @@
+        expect(section.find('.open-response-assessment-content').is(':visible')).toBe(true);
+
+        var expectedArr = [3, 3, 12, 3, 2, 0, 5, 2];
+        var arr = [];
```
changed this

dmitry-viskov commented:
Hi @efischer19

I've just sent to your email (efischer@edx.org) credentials for our Credo test server

dmitry-viskov commented:
@bjacobel there is no way to run `npm` during `pip install git+https://github.com/edx/edx-ora2.git@<hash>#egg=ora2==<version>`. So we have to add this backgrid dependency to the repository. If you want it could be done automatically in `make javascript` command: https://github.com/edx/edx-ora2/blob/master/Makefile#L32

dmitry-viskov commented:
hi @efischer19 . Thanks for the review!

>After this lands, the only remaining changes are in edx-platform, correct?

correct

dmitry-viskov commented:
@efischer19 @cahrens one little question. I've tested a lot of variants for ORA blocks but I can't reproduce such scenario when `staff` value in the _"Grade Available Responses"_ section differs from zero.
And as a consequence staff column in the grid in the instructor dashboard tab is also always zero.
May be I do something wrong but when student passes all steps except `staff` and the last step is `staff grade` system marks this learner as `waiting` instead of `staff` step. But after instructor assesses the students answer the new learners status is `done`
I may suggest that it is some kind of bug in ora block. What do you think?

![staff_zero](https://cloud.githubusercontent.com/assets/1876893/24067378/e682ac0a-0b8c-11e7-9ebc-0f04d1568c02.jpeg)
![staff_grid_field](https://cloud.githubusercontent.com/assets/1876893/24067379/ef53448e-0b8c-11e7-8bcc-d97955feed32.jpeg)

dmitry-viskov commented:
@efischer19 @cahrens I may suggest that [this statement](https://github.com/edx/edx-ora2/blob/master/openassessment/workflow/api.py#L303) is incorrect in case of `staff` status.
As for my opinion the correct query should be:
```
workflows = AssessmentWorkflow.objects.filter(course_id=course_id, item_id=item_id)
count = AssessmentWorkflowStep.objects.filter(workflow=workflows, assessment_completed_at=None, name='staff').count()
```

Am I wrong?

efischer19 commented:
@dmitry-viskov what you've noted is a "feature-not-bug" in ORA, related to the difference in what a student actively needs to do versus what needs to be done to the student's response that's out of their control. The steps listed will always be which action the student currently needs to undertake, which is "waiting" for students who need a staff grade. [See readthedocs for more](http://edx.readthedocs.io/projects/edx-partner-course-staff/en/latest/exercises_tools/open_response_assessments/Access_ORA_Info.html).

This has caused confusion before and may be worth changing in the future (FYI @sstack22, see [this previous JIRA discussion](https://openedx.atlassian.net/browse/TNL-4263?focusedCommentId=188464&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-188464) as well) but for the purposes of this PR it's expected behavior.

baldwin47 commented:
@efischer19 @sstack22 that's a deal breaker for us on this one - from our perspective, the entire purpose of a centralized list of ORA blocks is to make a place for staff members to come find which staff assessments are ready to grade. If the Staff value is always 0, then the value of this feature is very low (essentially useless for staff assessments). Adding in other ORA types was something we did based on edX's suggestion, in order to make this more likely to be merged to edx master. If you think this behavior may be worth changing in the future, why not just change it now?

efischer19 commented:
> from our perspective, the entire purpose of a centralized list of ORA blocks is to make a place for staff members to come find which staff assessments are ready to grade.

This is still a question that can be answered, it just needs to be asked slightly differently. "What states are the ORA responses in?" will yield the list of steps that we're currently discussing here - `self`, `peer`, `waiting`, or `done`. `staff` was never originally one of these steps, it was added later and has its own idiosyncracies (it's always available for overrides, there are never any student actions to complete, etc). This is why being in `staff` state does not imply an assessment is in need of staff grading and vice-versa. To find which responses currently need staff assessments, you would locate all responses in the `waiting` state on a problem that requires staff grading [as I've done on this commit somewhat sloppily](https://github.com/edx/edx-ora2/pull/986/commits/6b06ac20405795f38f846ac7a048606066de9450).

The reason we cannot change this now is that what we're discussing is a very low-level change - the statuses returned from `get_status_counts` are built from problem definitions, and we cannot partially redefine what the `staff` status means without viewing that entire framework. Similarly, the fix proposed by @dmitry-viskov above will not work, as *every* problem has a staff `AssessmentWorkflowStep` (this is what allows staff override scores to be made). The cleanest path we have forward on this PR is to accept that the underlying data (from `get_status_values`) does not have a perfect 1-to-1 mapping with the data we're trying to display, and add a translation layer similar to my work above in between the two to map `waiting` values to `needs staff grading` where appropriate.

Can you make this change? We're still waiting on a final accessibility review before we can merge this anyways. Essentially what you'll have is data from `collect_ora2_responses` as `{self: 2, staff: 0, waiting: 6}` that needs to be translated into `{self: 2, staff: 6}` for consumption by the front-end.

dmitry-viskov commented:
@efischer19 

>as I've done on this commit somewhat sloppily.

where should I look? :slightly_smiling_face:  are you sure that commit id is correct? :slightly_smiling_face: 

> Similarly, the fix proposed by @dmitry-viskov above will not work, as every problem has a staff AssessmentWorkflowStep

yes. I understand. I've suggested my fix only for problems that satisfy the condition: `"staff-assessment" in self.assessment_steps`. Is in this case my solution is not acceptable?

dmitry-viskov commented:
@efischer19 

ok. I see.

But why we can't use [such statement](https://github.com/edx/edx-ora2/blob/master/openassessment/assessment/models/staff.py#L66) for example:

```
ungraded = cls.objects.filter(
    models.Q(grading_started_at=None) | models.Q(grading_started_at__lte=timeout),
    course_id=course_id, item_id=item_id, grading_completed_at=None, cancelled_at=None
).count()

in_progress = cls.objects.filter(
    course_id=course_id, item_id=item_id, grading_completed_at=None, cancelled_at=None,
    grading_started_at__gt=timeout
).count()

graded = cls.objects.filter(
    course_id=course_id, item_id=item_id, cancelled_at=None
).exclude(grading_completed_at=None).count()

?

Or as I already told my first solution but only for the problems that satisfy the condition: `"staff-assessment" in self.assessment_steps`

?

efischer19 commented:
Sorry about that! Too many edits yesterday :man_facepalming: Here's a better link: https://github.com/edx/edx-ora2/pull/986/commits/c1ad84e2fa7bc3ed77018f68ebfb7df3616d6fc6#diff-8c9adfc3aff63db0a7bb063d710c9b80. Note that this is not changing the API behavior, it's simply throwing away the (always empty) `staff` status count, and treating `waiting` as the number we want in the case where problem requires staff grading.

As far as your proposed solutions - `"staff-assessment" in self.assessment_steps` will likely work, but we'll need to be sure to get `self.assessment_steps` from the *block* (the location in the course/the problem's raw definition) [using the property defined here](https://github.com/edx/edx-ora2/blob/76acce9523b9a1d4b54294872ae45930aacc6f44/openassessment/xblock/openassessmentblock.py#L630-L632) not the Workflow (per-submission django model) as [every Workflow has `'staff'` in `self.steps`](https://github.com/edx/edx-ora2/blob/master/openassessment/workflow/models.py#L388-L405). Your other suggestion using `StaffWorkflow` also might work, but I'm afraid we're digging in too deeply. I'd rather just take treat `waiting` statuses as `staff` iff `"staff-assessment" in self.assessment_steps`, else `peer`.

efischer19 commented:
That link is broken once again, here's a comment explaining all the commits on the other PR: https://github.com/edx/edx-ora2/pull/986#issuecomment-288109239. I'm not going to touch the history of the listed commits any further.

dmitry-viskov commented:
All changes are done

efischer19 commented:
Awesome, thanks @dmitry-viskov. Can you squash all these commits together, then we'll merge?

efischer19 commented:
From here:

1. I'll make a new edx-ora2 release, version 1.3.0
2. I'll update the base requirement in edx-platform as part of this process
3. Your edx-platform PR will need to be rebased to pick up this requirement update
4. edx-platform PR can merge

I'll begin reviewing the final PR while the ora release is ongoing; expect to see a message about the edx-platform ora2 version being updated sometime tomorrow.
